File: 2017/Nov2017/SolutionFindRestaurant.java
class SolutionFindRestaurant {
    public String[] findRestaurant(String[] list1, String[] list2) {
        Map<String, Integer> m = new HashMap<String, Integer>();
        for(int i = 0; i < list1.length; i++){
            m.put(list1[i], i);
        }
        List<String> tmp = new ArrayList<String>();
        int minSum = Integer.MAX_VALUE;
        for(int i = 0; i < list2.length; i++){
            if(m.containsKey(list2[i])){
                int tmpSum = m.get(list2[i]) + i;
                if(tmpSum == minSum){
                    tmp.add(list2[i]);
                }
                else if(tmpSum < minSum){
                    tmp.clear();
                    tmp.add(list2[i]);
                    minSum = tmpSum;
                }
            }
        }
        
        return tmp.toArray(new String[tmp.size()]);
        
    }
}

File: 2017/Nov2017/SolutionFindRestaurant.py
class SolutionFindRestaurant(object):
    def findRestaurant(self, list1, list2):
        """
        :type list1: List[str]
        :type list2: List[str]
        :rtype: List[str]
        """
        m  = {}
        tmp = []
        minSum = sys.maxsize
        for idx, name in enumerate(list1):
            m[name] = idx
        
        for idx, name in enumerate(list2):
            if name in m: 
                tmpSum = m[name] + idx;
                if(tmpSum == minSum):
                    tmp.append(name)
                if(tmpSum < minSum):
                    tmp = []
                    tmp.append(name)
                    minSum = tmpSum 
        return tmp
                    
        
        